File: Semester 5/PPD/proiect/examples/complex.py
"""
Complex example with solar system simulation and visualization.
"""

from shared.utils import generate_solar_system
from threaded.simulation import ThreadedNBodySimulation

try:
    from threaded.visualizer import NBodyVisualizer
    VISUALIZATION_AVAILABLE = True
except ImportError:
    VISUALIZATION_AVAILABLE = False
    print("Note: matplotlib not available for visualization")


def main():
    """Run a complex n-body simulation with visualization."""
    print("=" * 60)
    print("Solar System N-Body Simulation (Threaded)")
    print("=" * 60)
    
    # Generate solar system
    print("\nGenerating solar system...")
    bodies = generate_solar_system()
    
    print(f"Created {len(bodies)} celestial bodies:")
    for body in bodies:
        print(f"  - {body}")
    
    # Create simulation
    simulation = ThreadedNBodySimulation(bodies, num_threads=4)
    
    # Create visualizer if available
    visualizer = None
    if VISUALIZATION_AVAILABLE:
        visualizer = NBodyVisualizer()
        print("\nVisualization enabled")
    
    # Run simulation with periodic visualization updates
    print("\nRunning simulation...")
    num_steps = 1000
    dt = 3600.0  # 1 hour per step
    
    for step in range(num_steps):
        simulation.step(dt)
        
        if visualizer:
            visualizer.update(simulation.bodies)
        
        if (step + 1) % 100 == 0:
            kinetic, potential, total = simulation.get_energy()
            print(f"Step {step + 1}/{num_steps} | "
                  f"Time: {simulation.time_elapsed / 86400:.2f} days | "
                  f"Energy: {total:.2e} J")
    
    # Print final statistics
    print("\n" + "=" * 60)
    print("Final Statistics")
    print("=" * 60)
    stats = simulation.get_statistics()
    for key, value in stats.items():
        if key != 'center_of_mass':
            print(f"{key}: {value}")
    
    # Show visualization
    if visualizer:
        print("\nGenerating visualization...")
        visualizer.plot()
        visualizer.save("solar_system_trajectories.png")
        print("Visualization saved to solar_system_trajectories.png")
        # Uncomment to display interactively:
        # visualizer.show()


if __name__ == "__main__":
    main()

File: Semester 5/PPD/proiect/examples/complex.cpp
#include "threaded_simulation.h"
#include "utils.h"
#include <iostream>
#include <iomanip>

int main() {
    std::cout << "============================================================" << std::endl;
    std::cout << "Solar System N-Body Simulation (Threaded)" << std::endl;
    std::cout << "============================================================" << std::endl;
    
    // Generate solar system
    std::cout << "\nGenerating solar system..." << std::endl;
    auto bodies = generateSolarSystem();
    
    std::cout << "Created " << bodies.size() << " celestial bodies:" << std::endl;
    for (const auto& body : bodies) {
        std::cout << "  - " << body.toString() << std::endl;
    }
    
    // Create simulation
    ThreadedNBodySimulation simulation(bodies, 4);
    
    // Run simulation
    std::cout << "\nRunning simulation..." << std::endl;
    int num_steps = 1000;
    double dt = 3600.0;  // 1 hour per step
    
    for (int step = 0; step < num_steps; ++step) {
        simulation.step(dt);
        
        if ((step + 1) % 100 == 0) {
            auto [kinetic, potential, total] = simulation.getEnergy();
            std::cout << "Step " << (step + 1) << "/" << num_steps 
                      << " | Time: " << std::fixed << std::setprecision(2) 
                      << simulation.getStatistics().time_elapsed / 86400.0 << " days"
                      << " | Energy: " << std::scientific << total << " J" << std::endl;
        }
    }
    
    // Print final statistics
    std::cout << "\n============================================================" << std::endl;
    std::cout << "Final Statistics" << std::endl;
    std::cout << "============================================================" << std::endl;
    auto stats = simulation.getStatistics();
    std::cout << "num_bodies: " << stats.num_bodies << std::endl;
    std::cout << "num_threads: " << stats.num_threads << std::endl;
    std::cout << "time_elapsed: " << std::scientific << std::setprecision(2) 
              << stats.time_elapsed << std::endl;
    std::cout << "step_count: " << stats.step_count << std::endl;
    std::cout << "total_energy: " << stats.total_energy << std::endl;
    
    return 0;
}

